be/src/exec/pipeline/source_operator.h

```diff
@@ -55,13 +55,17 @@ class SourceOperatorFactory : public OperatorFactory {
     // - The scan operator, which has been assigned tablets with the specific bucket sequences.
     // - The exchange source operator, partitioned by HASH_PARTITIONED or BUCKET_SHUFFLE_HASH_PARTITIONED.
     virtual bool could_local_shuffle() const { return _could_local_shuffle; }
-    virtual void set_could_local_shuffle(bool could_local_shuffle) { _could_local_shuffle = could_local_shuffle; };
-    virtual TPartitionType::type partition_type() const { return TPartitionType::type::HASH_PARTITIONED; }
+    void set_could_local_shuffle(bool could_local_shuffle) { _could_local_shuffle = could_local_shuffle; }
+    virtual TPartitionType::type partition_type() const { return _partition_type; }
+    void set_partition_type(TPartitionType::type partition_type) { _partition_type = partition_type; };
+    virtual const std::vector<ExprContext*>& partition_exprs() const { return _empty_partition_exprs; }
 
 protected:
     size_t _degree_of_parallelism = 1;
     bool _could_local_shuffle = true;
+    TPartitionType::type _partition_type = TPartitionType::type::HASH_PARTITIONED;
     MorselQueueFactory* _morsel_queue_factory = nullptr;
+    const std::vector<ExprContext*> _empty_partition_exprs;
```

Comment:
it is weird, just in `virtual const std::vector<ExprContext*>& partition_exprs() const { return _empty_partition_exprs; }` 
```
return std::vector<ExprContext*>{};
```

Reply:
Cannot do this, because the return type is a const reference.